### PR TITLE
Feature/navigation

### PR DESCRIPTION
**Pull request by ChristianSteffensen**

### CODE REVIEW

File: amkotlinutil/build.gradle [hunk 1]
```diff
@@ -48,30 +49,36 @@ android {
 }
 
 dependencies {
-    compileOnly 'androidx.appcompat:appcompat:1.0.2'
-    compileOnly 'com.google.android.material:material:1.1.0-alpha07'
+    compileOnly 'androidx.appcompat:appcompat:1.1.0'
+    compileOnly 'com.google.android.material:material:1.1.0-alpha10'
     compileOnly "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
     compileOnly 'androidx.constraintlayout:constraintlayout:2.0.0-beta2'
```
Comment (martinloesethjensen) [inline amkotlinutil/build.gradle:52]: Don't know if we should downgrade this to beta1, since we had problems with it in other projects?

Comment (ChristianSteffensen) [reply amkotlinutil/build.gradle:52]: Good point 👍

File: amkotlinutil/build.gradle [hunk 2]
```diff
@@ -12,6 +12,7 @@ ext {
     playLibVersion = '10.0.1'
     conductorVersion = '2.1.5'
     firebaseVersion = '11.0.0'
+    rxBindingVersion = '3.0.0'
```
Comment (martinloesethjensen) [inline amkotlinutil/build.gradle]: We only use rxBindingVersion and conductorVersion

File: amkotlinutil/src/main/java/dk/adaptmobile/amkotlinutil/extensions/ConductorExtensions.kt [hunk 3]
```diff
@@ -159,3 +161,50 @@ var Controller.brightness: Float?
...
+}
+
+fun Router.goBack(amount: Int = 1, data: Any? = null) {
+    if (amount <= 1) {
+        popCurrentController()
+    } else {
+        val tempBackStack = backstack
+        for (i in 1..amount) {
+            tempBackStack.removeAt(tempBackStack.lastIndex)
+        }
```
Comment (martinloesethjensen) [inline amkotlinutil/src/main/java/dk/adaptmobile/amkotlinutil/extensions/ConductorExtensions.kt:181]: Where do we use `i`?

**Review by GaborHartyandi — APPROVED** [review]

File: amkotlinutil/src/main/java/dk/adaptmobile/amkotlinutil/navigation/BaseTab.kt [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package dk.adaptmobile.amkotlinutil.navigation
+
+import android.os.Bundle
+import androidx.annotation.IdRes
+
+interface BaseTabObject {
+    fun getInitialView() : BaseView<*, *>
+}
+
+abstract class BaseTab(@IdRes val id: Int, var state: Bundle? = null) : BaseTabObject
```
Comment (ChristianSteffensen) [inline amkotlinutil/src/main/java/dk/adaptmobile/amkotlinutil/navigation/BaseTab.kt]: Maybe we can remove BaseTapObject and instead pass in a function to the BaseTab constructor, which returns the view to initialize:
`abstract class BaseTab(@IdRes val id: Int, var initalView: () -> BaseView<*,*>, var state: Bundle? = null)`

Then the usage would change like this,which is cleaner imo. What do you think @martinloesethjensen 
 @GaborHartyandi 

**from**
```
sealed class Tab(id: Int, state: Bundle? = null) : BaseTab(id, state) {
    object Home : Tab(R.id.tab_home) {
        override fun getInitialView(): BaseView<*, *> = HomeView()
    }

    object Search : Tab(R.id.tab_search) {
        override fun getInitialView(): BaseView<*, *> = SearchView()
    }

    object Membercard : Tab(R.id.tab_membercard) {
        override fun getInitialView(): BaseView<*, *> {
            return if (PrefsManager.isUserLoggedIn) {
                MembercardView()
            } else {
                LoginView(LoginType.MemberCardTab)
            }
        }
    }

    object About : Tab(R.id.tab_about) {
        override fun getInitialView(): BaseView<*, *> = AboutView()
    }
}
```

**to:**
```
sealed class Tab(id: Int, state: Bundle? = null, initalView: () -> BaseView<*, *>) : BaseTab(id, initalView, state) {
    object Home : Tab(R.id.tab_home, {HomeView()})

    object Search : Tab(R.id.tab_search, {HomeView()})

    object Membercard : Tab(R.id.tab_membercard, {
        return if (PrefsManager.isUserLoggedIn) {
            MembercardView()
        } else {
            LoginView(LoginType.MemberCardTab)
        }
    })

    object About : Tab(R.id.tab_about, {HomeView()})
}
```

File: amkotlinutil/src/main/java/dk/adaptmobile/amkotlinutil/navigation/BaseView.kt [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+                            .doOnAndroidMain()
+                            .subscribe {
+//                                showLoadingBar(it)
+                            }
+
+                    viewModel.error
+                            .compose(bindUntilEvent(ControllerEvent.DESTROY))
+                            .doOnAndroidMain()
+                            .subscribe {
+                                //TODO: Handle error
```
Comment (martinloesethjensen) [inline amkotlinutil/src/main/java/dk/adaptmobile/amkotlinutil/navigation/BaseView.kt]: How should you handle the error?

File: amkotlinutil/src/main/java/dk/adaptmobile/amkotlinutil/navigation/BaseView.kt [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+            override fun postCreateView(controller: Controller, view: View) {
+                super.postCreateView(controller, view)
+                activity?.let { activity ->
+                    onViewBound(view, activity)
+
+                    viewModel.loading
+                            .compose(bindUntilEvent(ControllerEvent.DESTROY))
+                            .doOnAndroidMain()
+                            .subscribe {
+//                                showLoadingBar(it)
```
Comment (martinloesethjensen) [inline amkotlinutil/src/main/java/dk/adaptmobile/amkotlinutil/navigation/BaseView.kt]: Show a loading bar? 😅

File: amkotlinutil/src/main/java/dk/adaptmobile/amkotlinutil/navigation/BaseView.kt [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+                            }
+
+                    val application = activity.application as BaseApplicationController
+                    application.noNetworkSubject
+                            .compose(bindUntilEvent(ControllerEvent.DESTROY))
+                            .observeOn(AndroidSchedulers.mainThread())
+                            .subscribe({ hasNetwork ->
+                                           if (!hasNetwork) {
+                                               d(this.javaClass.name, "NoInternetDialog opens")
+//                                               NavManager.openModally(Routing.NoInternetDialog())
```
Comment (martinloesethjensen) [inline amkotlinutil/src/main/java/dk/adaptmobile/amkotlinutil/navigation/BaseView.kt]: Same as above ☝️

File: amkotlinutil/src/main/java/dk/adaptmobile/amkotlinutil/navigation/BaseView.kt [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+                                           }
+                                       }, {
+//                                           NavManager.openModally(Routing.NoInternetDialog())
```
Comment (martinloesethjensen) [inline amkotlinutil/src/main/java/dk/adaptmobile/amkotlinutil/navigation/BaseView.kt]: Same as above ☝️

File: amkotlinutil/src/main/java/dk/adaptmobile/amkotlinutil/extensions/ConductorExtensions.kt (hunk 3 above)
Comment (ChristianSteffensen) [reply amkotlinutil/src/main/java/dk/adaptmobile/amkotlinutil/extensions/ConductorExtensions.kt:181]: Nowhere, is there a way to iterate over a range without it?

File: amkotlinutil/src/main/java/dk/adaptmobile/amkotlinutil/navigation/BaseView.kt [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+    protected abstract fun handleOutput(output: T2)
+
+    open fun callback(data: Any?) {}
+
+    override fun onDestroyView(view: View) {
+        super.onDestroyView(view)
+        clearFindViewByIdCache()
+        viewModel.disposeBag.disposeSafe()
+    }
+
```
Comment (martinloesethjensen) [inline amkotlinutil/src/main/java/dk/adaptmobile/amkotlinutil/navigation/BaseView.kt:95]: This branch didn't have the ktlint? Maybe run ktlint after it's been merged to master 😉

**Review by martinloesethjensen — CHANGES_REQUESTED** [review]
Looks good - just some things that needs their logic moving. Like handle loading, error and no network subjects.

File: amkotlinutil/src/main/java/dk/adaptmobile/amkotlinutil/extensions/ConductorExtensions.kt [hunk 10]
```diff
@@ -159,3 +161,50 @@ var Controller.brightness: Float?
...
+fun Router.isEmpty(): Boolean {
+    return backstackSize == 0
+}
+
+fun Router.goBack(amount: Int = 1, data: Any? = null) {
+    if (amount <= 1) {
+        popCurrentController()
+    } else {
+        val tempBackStack = backstack
+        for (i in 1..amount) {
```
Comment (martinloesethjensen) [inline amkotlinutil/src/main/java/dk/adaptmobile/amkotlinutil/extensions/ConductorExtensions.kt:179]:
```suggestion
        (1..amount).forEach { tempBackStack.removeAt(tempBackStack.lastIndex) }
```

File: amkotlinutil/src/main/java/dk/adaptmobile/amkotlinutil/extensions/ConductorExtensions.kt [hunk 11]
```diff
@@ -159,3 +161,50 @@ var Controller.brightness: Float?
...
+            tempBackStack.removeAt(tempBackStack.lastIndex)
```
Comment (martinloesethjensen) [inline amkotlinutil/src/main/java/dk/adaptmobile/amkotlinutil/extensions/ConductorExtensions.kt:180]:
```suggestion
```

File: amkotlinutil/src/main/java/dk/adaptmobile/amkotlinutil/extensions/ConductorExtensions.kt (hunk 3 above)
Comment (martinloesethjensen) [inline amkotlinutil/src/main/java/dk/adaptmobile/amkotlinutil/extensions/ConductorExtensions.kt:181]:
```suggestion
```

Comment (martinloesethjensen) [reply amkotlinutil/src/main/java/dk/adaptmobile/amkotlinutil/extensions/ConductorExtensions.kt:181]: Do a foreach without index ;)

**Review by martinloesethjensen — CHANGES_REQUESTED** [review]
Maybe a foreach without index

**Pull request merged**